### PR TITLE
CASSANDRA-18212: Reduce memory allocations of calls to ByteBufer.duplicate() made in org.apache.cassandra.transport.CBUtil#writeValue

### PR DESCRIPTION
**Pull request by NateAdere**
Used HeapByteBuffers and ThreadLocal to allocate a ByteBuffer for every thread. 
 * Reduced 20% memory allocation when doing mostly memory based queries
 * Reduced 4% memory allocation when doing disk-based queries
  
```
patch by Natnael Adere; reviewed by <Reviewers> for CASSANDRA-#####
```

The [Cassandra Jira](https://issues.apache.org/jira/browse/CASSANDRA-18212)

### CODE REVIEW

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    private void directTest(int capacity, boolean readOnly)
+    {
+        ByteBuffer hollowBuffer = CBUtil.getLocalBuffer();
+        ByteBuffer bb = readOnly ? ByteBuffer.allocateDirect(capacity).asReadOnlyBuffer() : ByteBuffer.allocateDirect(capacity);
+
+        int size = bb.capacity();
+        buf = allocator.directBuffer(size);
+
+        try {
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: braces on new line in this whole try catch

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    private void heapTest(int capacity, boolean readOnly)
+    {
+        ByteBuffer hollowBuffer = CBUtil.getLocalBuffer();
+        ByteBuffer bb = readOnly ? ByteBuffer.allocate(capacity).asReadOnlyBuffer() : ByteBuffer.allocate(capacity);
+
+        int size = bb.capacity();
+        buf = allocator.heapBuffer(size);
+
+        try {
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: same

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Test
+    public void CBUtilWriteValueTest()
+    {
+        ByteBuffer bb = ByteBuffer.allocate(50);
+        int size = bb.capacity();
+        buf = allocator.heapBuffer(size);
+
+        CBUtil.writeValue(bb, buf);
+        Assert.assertEquals(0, buf.readerIndex());
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: cant you import it statically? saves some time to read it

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        ByteBuffer bb = readOnly ? ByteBuffer.allocateDirect(capacity).asReadOnlyBuffer() : ByteBuffer.allocateDirect(capacity);
+
+        int size = bb.capacity();
+        buf = allocator.directBuffer(size);
+
+        try {
+            MemoryUtil.duplicateHeapByteBuffer(bb, hollowBuffer);
+        }
+        catch (Error e) {
+            Assert.assertEquals(new AssertionError().getClass(), e.getClass());
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: why do you need to create an instance of AssertionError to call getClass() on it? Does not `AssertionError.class.getClass()` do the job?

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        try {
+            ByteBuffer temp = MemoryUtil.duplicateHeapByteBuffer(bb, hollowBuffer);
+            Assert.assertEquals(bb.position(), temp.position());
+            Assert.assertEquals(bb.limit(), temp.limit());
+            Assert.assertEquals(bb.capacity(), temp.capacity());
+            Assert.assertEquals(bb.array(), temp.array());
+        }
+        catch (Exception e){
+            Assert.assertEquals(new ReadOnlyBufferException().getClass(), e.getClass());
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: `ReadOnlyBufferException.class.getClass()` maybe?

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 6]
```diff
@@ -46,7 +46,7 @@
 import org.apache.cassandra.utils.Pair;
 import org.apache.cassandra.utils.TimeUUID;
 import org.apache.cassandra.utils.UUIDGen;
-
+import org.apache.cassandra.utils.memory.MemoryUtil;
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/transport/CBUtil.java:48]: should not be there some free line after imports and before the beginning of comments? This feels strange.

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 7]
```diff
@@ -614,4 +623,9 @@ private static byte[] readRawBytes(ByteBuf cb, int length)
         return bytes;
     }
 
+    public static ByteBuffer getLocalBuffer()
+    {
+        return localBuffer.get();
+    }
+
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: empty line probably redundant?

File: test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
+import org.apache.cassandra.utils.memory.MemoryUtil;
+
+public class DuplicateHeapBufferTest
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/transport/DuplicateHeapBufferTest.java]: spoke already, we should remove as you moved away from heap buffer copying

File: src/java/org/apache/cassandra/utils/memory/MemoryUtil.java [hunk 9]
```diff
@@ -34,19 +34,20 @@
 
     private static final Unsafe unsafe;
     private static final Class<?> DIRECT_BYTE_BUFFER_CLASS, RO_DIRECT_BYTE_BUFFER_CLASS;
-    private static final long DIRECT_BYTE_BUFFER_ADDRESS_OFFSET;
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/utils/memory/MemoryUtil.java]: can you revert all changes?  as you migrated away from the heap buffer duplicate logic this logic is no longer needed to be changed

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 10]
```diff
@@ -614,4 +653,9 @@ private static byte[] readRawBytes(ByteBuf cb, int length)
         return bytes;
     }
 
+    public static ByteBuffer getLocalBuffer()
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: I would leave private for now

Comment (dcapwell) [reply src/java/org/apache/cassandra/transport/CBUtil.java]: also can you rename to `getLocalDirectBuffer`?  so callers are more clear

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 11]
```diff
@@ -478,7 +487,37 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
         cb.writeInt(remaining);
 
         if (remaining > 0)
-            cb.writeBytes(bytes.duplicate());
+        {
+            cb.writeBytes(MemoryUtil.duplicateHeapByteBuffer(bytes, getLocalBuffer()));
+        }
+    }
+
+    public static void writeValue(ByteBuffer src, ByteBuf dest)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: should not copy/paste an existing method...

You "should" add a new method that can write a `ByteBuffer` to a `ByteBuf` without allocating, then have the one line `cb.writeBytes(bytes.duplicate());` call that method instead.

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 12]
```diff
@@ -466,7 +475,7 @@ public static void writeValue(byte[] bytes, ByteBuf cb)
         cb.writeBytes(bytes);
     }
 
-    public static void writeValue(ByteBuffer bytes, ByteBuf cb)
+    public static void writeValu(ByteBuffer bytes, ByteBuf cb)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: please revert

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 13]
```diff
@@ -69,6 +69,15 @@ protected CharsetDecoder initialValue()
         }
     };
 
+    private final static FastThreadLocal<ByteBuffer> localBuffer = new FastThreadLocal<ByteBuffer>()
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: would be best to call it `localDirectBuffer`

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 14]
```diff
@@ -478,7 +487,37 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
...
+            dest.writeInt(-1);
+            return;
+        }
+
+        int length = src.remaining();
+        dest.writeInt(length);
+
+        if (src.hasArray())
+        {
+            byte[] array = src.array();
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java:506]: can you add a comment explaining why we do this when direct duplicates the buffer?  You found a JDK bug, we should expose that to anyone reading so they know that there may be dragons to replicate

**Review by dcapwell — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 15]
```diff
@@ -478,7 +487,35 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
         cb.writeInt(remaining);
 
         if (remaining > 0)
-            cb.writeBytes(bytes.duplicate());
+        {
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: remove the `{}`

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 16]
```diff
@@ -478,7 +487,35 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
...
+            addBytes(bytes, cb);
+        }
+    }
+
+    public static void addBytes(ByteBuffer src, ByteBuf dest)
+    {
+        if (src == null)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: why the null check?  at least in the single usage this doesn't allow null... I prefer if we remove and allow the NPE if a user provides null

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 17]
```diff
@@ -299,6 +299,16 @@ private static char[] createDNSDomainPartDomain()
...
+    {
+        return bytes(min, max, SourceDSL.arbitrary().constant(BBCases.HEAP));
+    }
+
+    public static Gen<ByteBuffer> bytesAnyType(int min, int max)
+    {
+        return bytes(min, max, SourceDSL.arbitrary().enumValues(BBCases.class));
+    }
+
+    public static Gen<ByteBuffer> bytes(int min, int max, Gen<BBCases> cases)
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: have to make this private, or fix `BBCases`.  You can't have a public method expose a private class (the build should be failing right now...).

If you choose to keep this method (I would not) then you should rename `BBCases` to be more clear to readers...

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 18]
```diff
@@ -314,11 +324,32 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+            switch (cases.generate(rnd))
+            {
+                case HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size);
+                case READ_ONLY_HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size).asReadOnlyBuffer();
+                case DIRECT:
+                {
+                    ByteBuffer bb = ByteBuffer.allocateDirect(size);
+                    bb.put(LazySharedBlob.SHARED_BYTES, offset, size);
+                    bb.flip();
+                    return bb;
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: can you make this a function?  best not to copy/paste

File: test/unit/org/apache/cassandra/transport/WriteBytesDuplicateTest.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.apache.cassandra.utils.Generators;
+import org.assertj.core.api.Assertions;
+
+import static org.quicktheories.QuickTheory.qt;
+
+
+public class WriteBytesDuplicateTest
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/transport/WriteBytesDuplicateTest.java]: why `Duplicate`?  You are "writing"

**Review by dcapwell — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/transport/CBUtil.java (hunk 6 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/transport/CBUtil.java:48]: @NateAdere  why was this comment resolved?

File: src/java/org/apache/cassandra/transport/CBUtil.java [hunk 20]
```diff
@@ -478,7 +487,32 @@ public static void writeValue(ByteBuffer bytes, ByteBuf cb)
...
+    }
+
+    public static void addBytes(ByteBuffer src, ByteBuf dest)
+    {
+        if (src.remaining() == 0)
+            return;
+
+        int length = src.remaining();
+
+        //heap buffers are copied this way in order to avoid JVM crashing
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/transport/CBUtil.java]: please give more detail, this isn't enough for anyone to understand why this code is this way.  You should also move it within the `src.hasArray` block as that comment only makes sense there.

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 21]
```diff
@@ -314,11 +324,36 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+        switch (cases.generate(rnd))
+        {
+            case HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size);
+            case READ_ONLY_HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size).asReadOnlyBuffer();
+            case DIRECT:
+            {
+                ByteBuffer bb = ByteBuffer.allocateDirect(size);
+                bb.put(LazySharedBlob.SHARED_BYTES, offset, size);
+                bb.flip();
+                return bb;
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: when I said you should avoid copy/paste and make this a method, I mean 

```
ByteBuffer bb = ByteBuffer.allocateDirect(size);
                 bb.put(LazySharedBlob.SHARED_BYTES, offset, size);
                 bb.flip();
```

and not this new `handleCases`...

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 22]
```diff
@@ -314,11 +324,36 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+                bb.put(LazySharedBlob.SHARED_BYTES, offset, size);
+                bb.flip();
+                return bb.asReadOnlyBuffer();
+            }
+            default: throw new AssertionError("cann't wait for jdk 17!");
+        }
     }
 
-    /**
+     /**
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java:351]: please revert

**Review by dcapwell — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/transport/CBUtil.java (hunk 14 above)
Comment (NateAdere) [reply src/java/org/apache/cassandra/transport/CBUtil.java:506]: Is this comment descriptive enough?

File: test/unit/org/apache/cassandra/utils/Generators.java (hunk 21 above)
Comment (NateAdere) [reply test/unit/org/apache/cassandra/utils/Generators.java]: the new handleCases helps with readability so my plan is to keep that method and add "ByteBuffer bb = directBufferFromSharedBlob(offset, size);" to address the copy/paste code. Any opinions on removing or keeping the method?

File: src/java/org/apache/cassandra/transport/CBUtil.java (hunk 14 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/transport/CBUtil.java:506]: can you flesh this out more?  Scott showed you the bug.

it is also not an "improper" way, it was a bug in CMS

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 23]
```diff
@@ -314,11 +324,39 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+    private static ByteBuffer handleCases(Gen<BBCases> cases, RandomnessSource rnd, int offset, int size) {
+        switch (cases.generate(rnd))
+        {
+            case HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size);
+            case READ_ONLY_HEAP: return ByteBuffer.wrap(LazySharedBlob.SHARED_BYTES, offset, size).asReadOnlyBuffer();
+            case DIRECT:
+            {
+                ByteBuffer bb = directBufferFromSharedBlob(offset, size);
+                return bb;
+            }
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: nit: cleanup, you can just do `return directBufferFromSharedBlob(offset, size);` which lets you get rid of 3 lines of code

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 24]
```diff
@@ -314,11 +324,39 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+            case READ_ONLY_DIRECT:
+            {
+                ByteBuffer bb = directBufferFromSharedBlob(offset, size);
+                return bb.asReadOnlyBuffer();
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: nit: cleanup, you can just do `return directBufferFromSharedBlob(offset, size);` which lets you get rid of 3 lines of code

File: test/unit/org/apache/cassandra/utils/Generators.java [hunk 25]
```diff
@@ -314,11 +324,39 @@ public static Gen<ByteBuffer> bytes(int min, int max)
...
+            }
+            default: throw new AssertionError("cann't wait for jdk 17!");
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/utils/Generators.java]: fix typo plz

**Review by dcapwell — APPROVED** [review]

**Comment by michaelsembwever** [comment]
Committed with https://github.com/apache/cassandra/commit/35f8da66f9b05d18b4177f8d2e1b86c772ad2221

**Pull request closed without merging**